File: backend/services/probability.py
from datetime import datetime

SLEEP_HOURS_PER_DAY = 8
FREE_FRACTION = (24 - SLEEP_HOURS_PER_DAY) / 24  # 16/24


def free_hours(deadline_iso: str) -> float:
    deadline = datetime.fromisoformat(deadline_iso)
    hours_raw = (deadline - datetime.now()).total_seconds() / 3600
    return max(0.0, hours_raw * FREE_FRACTION)


def calculate_probabilities(tasks: list[dict]) -> dict:
    """
    P(i) = 1 - E_i / T_left_i
    T_left_i = F_i - sum(j!=i) [ P_j * E_j * (F_i - E_j/2) / F_i ]
    F_i = free hours before deadline (16/24 of raw time)
    """
    eligible = [
        t for t in tasks
        if t.get("status") == "pending"
        and t.get("deadline")
        and t.get("estimated_hours", 0) > 0
    ]

    if not eligible:
        return {}

    F = {t["id"]: free_hours(t["deadline"]) for t in eligible}

    # Initial naive estimate
    P = {
        t["id"]: min(1.0, max(0.0, 1 - t["estimated_hours"] / F[t["id"]]))
        if F[t["id"]] > 0 else 0.0
        for t in eligible
    }

    for _ in range(30):
        max_diff = 0.0
        new_P = {}
        for ti in eligible:
            Fi = F[ti["id"]]
            Ei = ti["estimated_hours"]
            if Fi <= 0:
                new_P[ti["id"]] = 0.0
                continue
            stolen = sum(
                P[tj["id"]] * tj["estimated_hours"] * (Fi - tj["estimated_hours"] / 2) / Fi
                for tj in eligible
                if tj["id"] != ti["id"]
                and max(0.0, P[tj["id"]] * tj["estimated_hours"] * (Fi - tj["estimated_hours"] / 2) / Fi) > 0
            )
            Tleft = max(0.0, Fi - stolen)
            new_p = min(1.0, max(0.0, 1 - Ei / Tleft)) if Tleft > 0 else 0.0
            max_diff = max(max_diff, abs(new_p - P[ti["id"]]))
            new_P[ti["id"]] = new_p
        P = new_P
        if max_diff < 0.001:
            break

    return P


def get_bucket(p: float) -> str:
    if p > 0.80: return "safe"
    if p > 0.60: return "watch"
    if p > 0.40: return "at-risk"
    if p > 0.20: return "danger"
    return "critical"

File: frontend/src/utils/probability.js
const SLEEP_HOURS_PER_DAY = 8
const FREE_FRACTION = (24 - SLEEP_HOURS_PER_DAY) / 24  // 16/24

export function freeHours(deadlineIso) {
  const hoursRaw = (new Date(deadlineIso) - Date.now()) / 3_600_000
  return Math.max(0, hoursRaw * FREE_FRACTION)
}

export function calculateProbabilities(tasks) {
  // Only pending tasks with both deadline and estimated_hours
  const eligible = tasks.filter(
    t => t.status === 'pending' && t.deadline && t.estimated_hours > 0
  )

  if (eligible.length === 0) return {}

  const F = {}
  eligible.forEach(t => { F[t.id] = freeHours(t.deadline) })

  // Initial estimate: naive single-task probability
  const P = {}
  eligible.forEach(t => {
    P[t.id] = F[t.id] > 0 ? Math.min(1, Math.max(0, 1 - t.estimated_hours / F[t.id])) : 0
  })

  // Iterate to convergence
  for (let iter = 0; iter < 30; iter++) {
    let maxDiff = 0

    eligible.forEach(ti => {
      const Fi = F[ti.id]
      const Ei = ti.estimated_hours
      if (Fi <= 0) { P[ti.id] = 0; return }

      let stolen = 0
      eligible.forEach(tj => {
        if (tj.id === ti.id) return
        const Pj = P[tj.id]
        const Ej = tj.estimated_hours
        // Closed-form: E[time tj steals from ti's window]
        const contribution = Pj * Ej * (Fi - Ej / 2) / Fi
        stolen += Math.max(0, contribution)
      })

      const Tleft = Math.max(0, Fi - stolen)
      const newP = Ei > 0 && Tleft > 0
        ? Math.min(1, Math.max(0, 1 - Ei / Tleft))
        : (Tleft <= 0 ? 0 : 1)

      maxDiff = Math.max(maxDiff, Math.abs(newP - P[ti.id]))
      P[ti.id] = newP
    })

    if (maxDiff < 0.001) break
  }

  return P  // { [task_id]: probability }
}

export function getBucket(p) {
  if (p === null || p === undefined) return null
  if (p > 0.80) return 'safe'
  if (p > 0.60) return 'watch'
  if (p > 0.40) return 'at-risk'
  if (p > 0.20) return 'danger'
  return 'critical'
}

export const BUCKET_COLOR = {
  safe:     '#38a169',
  watch:    '#68d391',
  'at-risk':'#d69e2e',
  danger:   '#ed8936',
  critical: '#e53e3e',
}

const BUCKET_ORDER = ['safe', 'watch', 'at-risk', 'danger', 'critical']

export function bucketDropped(prev, next) {
  if (!prev || !next) return false
  return BUCKET_ORDER.indexOf(next) > BUCKET_ORDER.indexOf(prev)
}
